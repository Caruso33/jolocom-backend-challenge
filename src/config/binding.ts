import { AsyncContainerModule, Container } from 'inversify'
import { TYPES } from '../types'
import {} from '../controller'
import { HelloController } from '../hello/helloController'
import { Greeter } from '../hello/greeter'

export const binding = new AsyncContainerModule(async (bind) => {
  // you might want to create your db connection here and setup your repositories for injection

  //World
  bind(TYPES.Controller).to(HelloController)
  bind<Greeter>(Greeter).toSelf().inSingletonScope()
})

export const createAppContainer = async (): Promise<Container> => {
  const container = new Container()
  await container.loadAsync(binding)
  return container
}
