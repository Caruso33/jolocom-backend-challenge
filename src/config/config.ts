import appRootDir from 'app-root-path'
import dotenv from 'dotenv'
import { strict as assert } from 'assert'
import { Oas3ToolsObjectOrientedConfig } from '@jolocom/oas3-tools-object-oriented'
import { oas3AppOptions } from './oas3'
import { OpenApiRequestHandler } from 'express-openapi-validator/dist/framework/types'

// Defining default environment variables
process.env.NODE_ENV =
  process.env.NODE_ENV === 'test'
    ? process.env.NODE_ENV
    : process.env.APP_ENV || 'dev'
process.env.API_VERSION = process.env.API_VERSION || 'v1'

// Loading env variables defined in .env file into the process.env scope
const envFound = dotenv.config()

if (envFound.error) {
  throw new Error("Couldn't find .env file")
}

// Asserting that required env variables are defined
const assertRequiredVariablesDefined = () => {
  assert(process.env.APP_PORT, "Expected 'process.env.APP_PORT' to be defined")
  assert(process.env.APP_ENV, "Expected 'process.env.APP_ENV' to be defined")
  assert(process.env.APP_HOST, "Expected 'process.env.APP_HOST' to be defined")
  assert(
    process.env.APP_SCHEME,
    "Expected 'process.env.APP_SCHEME' to be defined",
  )
  assert(
    process.env.SDK_CALLBACK_URL,
    "Expected 'process.env.SDK_CALLBACK_URL' to be defined",
  )
}

assertRequiredVariablesDefined()

/**
 * The main application configuration definition abstraction.
 */
export interface AppConfig {
  port: string
  env: string
  host: string
  schema: string
  apiVersion: string
  apiRootPath: string
  controllersPath: string

  logger: {
    logDir: string
  }

  openapi: (
    customMiddlewares?: OpenApiRequestHandler[],
  ) => Oas3ToolsObjectOrientedConfig
}

const oas3DeclarationFilePath = appRootDir + '/src/api/openapi.yaml'

/**
 * The main application configuration definition.
 */
export const config: AppConfig = {
  port: process.env.APP_PORT as string,
  env: process.env.APP_ENV as string,
  host: process.env.APP_HOST as string,
  schema: process.env.APP_SCHEME as string,
  apiVersion: process.env.API_VERSION as string,
  apiRootPath: '/api/' + process.env.API_VERSION,
  controllersPath: appRootDir + '/src/controller',

  logger: {
    logDir: process.env.LOG_DIR || appRootDir + '/var/log',
  },

  openapi: (customMiddlewares?: OpenApiRequestHandler[]) => ({
    oas3AppOptions: oas3AppOptions(oas3DeclarationFilePath),
    oas3DeclarationFilePath,
    customMiddlewares,
  }),
}
