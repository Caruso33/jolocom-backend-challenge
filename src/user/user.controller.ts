import { Response } from 'express'
import { controller, httpGet, httpPost } from 'inversify-express-utils'
import { CreateUser, User, ErrorMessage } from '../api/generated-schemas/models'
import { AppRequest } from '../middleware/appRequest'
import { UserService } from './user.service'

// @injectable()
@controller('/user')
export class UserController {
  constructor(private readonly user: UserService) {}

  /**
   * Create an user
   *
   * @param request create request with a body of user data
   * @param response response containing a {@link User}
   * @return {Promise<User>}
   */
  @httpPost('/')
  public async createUser(
    request: AppRequest<CreateUser>,
    response: Response<User>,
  ) {
    const createdUser = await this.user.createUser(request.body)

    return response.status(201).json(createdUser)
  }

  /**
   * Get an user
   *
   * @param request get request with an id parameter and no body
   * @param response response containing a {@link User}
   * @return {Promise<User | ErrorMessage>}
   */
  @httpGet('/:id')
  public async getUser(
    request: AppRequest<undefined, { id: string }>,
    response: Response<User | ErrorMessage>,
  ) {
    const user = await this.user.getUser(+request.pathParams.id)

    if (!user) {
      return response.status(404).json({ message: 'User not found' })
    }

    return response.status(200).json(user)
  }
}
