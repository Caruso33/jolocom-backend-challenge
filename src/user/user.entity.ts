import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm'

@Entity()
export class UserMetaEntity {
  @PrimaryGeneratedColumn()
  id: number

  @Column({ default: false })
  hasJoinedInvitation: boolean

  @CreateDateColumn()
  createdAt: Date

  @UpdateDateColumn()
  updatedAt: Date
}

@Entity()
export class UserEntity {
  @PrimaryGeneratedColumn()
  id: number

  @Column()
  name: string

  @Column()
  email: string

  @OneToOne(() => UserMetaEntity)
  @JoinColumn()
  metadata: UserMetaEntity
}
