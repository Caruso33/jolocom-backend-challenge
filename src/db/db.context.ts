import { injectable } from 'inversify'
import { DataSource } from 'typeorm'
import { UserEntity, UserMetaEntity } from '../user/user.entity'
import 'dotenv/config'

export function getDatabase(env: string) {
  const envDbMapping: Record<string, string> = {
    test: ':memory:',
    dev: './jolocom.sqlite',
  }

  return envDbMapping[env]
}

export function getDataSource(getOriginalDatabase = false) {
  const database = getDatabase(
    getOriginalDatabase || !process.env.NODE_ENV
      ? (process.env.APP_ENV as string)
      : (process.env.NODE_ENV as string),
  )

  const datasource = new DataSource({
    type: 'sqlite',
    database,
    synchronize: false,
    logging: false,
    entities: [__dirname + '/../**/*.entity.{js,ts}'],
    migrations: [__dirname + '/../**/migration/*.{js,ts}'],
    subscribers: [],
  })

  return datasource
}

export const datasource = getDataSource()

@injectable()
export class DBContext {
  private _db: DataSource

  async init() {
    this._db = await getDataSource().initialize()

    console.log('connected to DB')
  }

  async runMigrations() {
    await this._db.runMigrations()

    console.log('ran migrations')
  }

  async undoLastMigration() {
    await this._db.undoLastMigration()

    console.log('undid last migration')
  }

  async showMigrations() {
    return await this._db.showMigrations()
  }

  async close() {
    this._db.destroy()
  }

  get user() {
    return this._db.getRepository(UserEntity)
  }

  get userMeta() {
    return this._db.getRepository(UserMetaEntity)
  }
}
