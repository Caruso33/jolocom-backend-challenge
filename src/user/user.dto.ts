import { User } from '../api/generated-schemas/models'
import { UserEntity } from './user.entity'

export class UserDto {
  // converts user entities from db to dtos for api transfer
  static to(user: UserEntity) {
    const dto: User = {
      ...user,
      hasJoinedInvitation: user.hasJoinedInvitation,
      createdAt: user.createdAt.getTime(),
      updatedAt: user.updatedAt.getTime(),
    }

    return dto
  }
}
