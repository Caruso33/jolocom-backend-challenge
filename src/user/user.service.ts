import { injectable } from 'inversify'
import { CreateUser } from '../api/generated-schemas/models'
import { UserDto } from './user.dto'
import { UserEntity, UserMetaEntity } from './user.entity'
import { UserMetaRepository, UserRepository } from './user.repository'

/* 
  Business logic of user 
*/
@injectable()
export class UserService {
  constructor(
    private readonly userRepo: UserRepository,
    private readonly userMetaRepo: UserMetaRepository,
  ) {}

  async createUser(user: CreateUser) {
    // basic validation

    if (!user.name || typeof user.name !== 'string') {
      throw Error(`User name is missing`)
    }

    const emailRegex =
      /^[a-zA-Z0-9]+(?:\.[a-zA-Z0-9]+)*@[a-zA-Z0-9]+(?:\.[a-zA-Z0-9]+)*$/

    if (
      !user.email ||
      typeof user.email !== 'string' ||
      !emailRegex.test(user.email)
    ) {
      throw Error(`User email is missing or has wrong format`)
    }

    if (!user.hasJoinedInvitation) {
      user.hasJoinedInvitation = false
    }

    const newMetaUser = new UserMetaEntity()
    newMetaUser.hasJoinedInvitation = user.hasJoinedInvitation

    await this.userMetaRepo.create(newMetaUser)

    const newUser = new UserEntity()
    newUser.name = user.name
    newUser.email = user.email
    newUser.metadata = newMetaUser

    await this.userRepo.create(newUser)

    return UserDto.toUserEntity(newUser)
  }

  async getUser(id: UserEntity['id']) {
    const user = await this.userRepo.findOne(id)

    if (!user) {
      return null
    }

    return UserDto.toUserEntity(user)
  }
}
