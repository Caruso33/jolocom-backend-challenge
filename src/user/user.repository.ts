import { injectable } from 'inversify'
import { CreateUser } from '../api/generated-schemas/models'
import { DBContext } from '../db/db.context'
import { UserEntity } from './user.entity'

/* 
  Lower level repository to interact with database layer
*/
@injectable()
export class UserRepository {
  constructor(private readonly _dbContext: DBContext) {}

  async create(user: CreateUser) {
    return this._dbContext.user.save(user)
  }

  async findOne(id: UserEntity['id']) {
    return this._dbContext.user.findOne({ where: { id } })
  }
}
