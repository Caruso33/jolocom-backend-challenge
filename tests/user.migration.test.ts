import { Application } from 'express'
import fs from 'fs'
import { Container } from 'inversify'
import 'reflect-metadata'
import createApp from '../src/app'
import { createAppContainer } from '../src/config/binding'
import { config } from '../src/config/config'
import { DBContext, getDataSource } from '../src/db/db.context'
import { UserService } from './../src/user/user.service'
import { getTestUser } from './utils'

describe('User Test', () => {
  let container: Container, dbContext: DBContext
  let isMigrationTestNeeded: boolean

  beforeAll(async () => {
    const appContainer = await createAppContainer()

    const createdApp: [Application, Container] = await createApp(
      config,
      appContainer,
    )
    container = createdApp[1]

    dbContext = await container.get(DBContext)
    await dbContext.runMigrations()

    const origDbContext = await getDataSource(true).initialize()
    isMigrationTestNeeded = await origDbContext.showMigrations()
    await origDbContext.destroy()
  })

  afterAll(async () => {
    const dbContext = await container.get(DBContext)
    await dbContext.close()
  })

  it('ensures creating an user is backwards compatible', async () => {
    const user = getTestUser()
    const userService: UserService = await container.get(UserService)

    if (!isMigrationTestNeeded) {
      console.log(`No migration tests needed`)

      if (!fs.existsSync('./__snapshots__/user.migration.test.ts.snap')) {
        console.log(
          `Setting up initial snapshot to compare user.service returns against in the future`,
        )
        await userService.createUser(user)
        const foundUser = await userService.getUser(1)
        expect(foundUser).toMatchSnapshot({
          createdAt: expect.any(Number),
          updatedAt: expect.any(Number),
        })
      }
      return
    }

    await dbContext.undoLastMigration()

    let users = await dbContext.user.query(
      `SELECT "id", "name", "email", "hasJoinedInvitation", "createdAt", "updatedAt" FROM "user_entity" LIMIT 1`,
    )
    expect(users.length).toBe(0)

    await dbContext.user.query(
      `INSERT INTO "user_entity" ("name", "email", "hasJoinedInvitation") VALUES ("${user.name}", "${user.email}", ${user.hasJoinedInvitation})`,
    )

    users = await dbContext.user.query(
      `SELECT "id", "name", "email", "hasJoinedInvitation", "createdAt", "updatedAt" FROM "user_entity" LIMIT 1`,
    )
    expect(users.length).toBe(1)
    expect(users[0]).toMatchObject({ ...user, hasJoinedInvitation: 1 }) // bools are saved as `0` / `1`

    await dbContext.runMigrations()

    const foundUser = await userService.getUser(1)

    expect(foundUser).toMatchObject({
      id: users[0].id,
      name: users[0].name,
      email: users[0].email,
      hasJoinedInvitation: !!users[0].hasJoinedInvitation, // see above: bools are saved as `0` / `1`
    })
    expect(foundUser).toHaveProperty('createdAt')
    expect(foundUser?.createdAt).toBeGreaterThan(0)
    expect(foundUser).toHaveProperty('updatedAt')
    expect(foundUser?.updatedAt).toBeGreaterThan(0)

    // testing snapshot against one from before migration
    expect(foundUser).toMatchSnapshot({
      createdAt: expect.any(Number),
      updatedAt: expect.any(Number),
    })
  })
})
