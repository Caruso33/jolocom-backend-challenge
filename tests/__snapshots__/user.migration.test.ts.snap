// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Test ensures creating an user is backwards compatible 1`] = `
Object {
  "createdAt": Any<Number>,
  "email": "elrond@lindon.cc",
  "hasJoinedInvitation": true,
  "id": 1,
  "name": "elrond",
  "updatedAt": Any<Number>,
}
`;
